test/extensions/filters/udp/udp_proxy/session_filters/dynamic_forward_proxy/proxy_filter_test.cc

```diff
@@ -0,0 +1,311 @@
+  EXPECT_CALL(*dns_cache_manager_->dns_cache_, loadDnsCacheEntry_(Eq("host"), 50, _, _))
+      .WillOnce(Return(
+          MockLoadDnsCacheEntryResult{LoadDnsCacheEntryStatus::InCache, nullptr, absl::nullopt}));
+  EXPECT_EQ(ReadFilterStatus::Continue, filter_->onNewSession());
+  EXPECT_EQ(ReadFilterStatus::Continue, filter_->onData(recv_data_stub1_));
+}
+
+TEST_F(DynamicProxyFilterTest, LoadingCacheEntry) {
+  setup();
+  setFilterState("host", 50);
+  Upstream::ResourceAutoIncDec* circuit_breakers_{
+      new Upstream::ResourceAutoIncDec(pending_requests_)};
+  EXPECT_CALL(*dns_cache_manager_->dns_cache_, canCreateDnsRequest_())
+      .WillOnce(Return(circuit_breakers_));
+  Extensions::Common::DynamicForwardProxy::MockLoadDnsCacheEntryHandle* handle =
+      new Extensions::Common::DynamicForwardProxy::MockLoadDnsCacheEntryHandle();
+  EXPECT_CALL(*dns_cache_manager_->dns_cache_, loadDnsCacheEntry_(Eq("host"), 50, _, _))
+      .WillOnce(Return(
+          MockLoadDnsCacheEntryResult{LoadDnsCacheEntryStatus::Loading, handle, absl::nullopt}));
+  EXPECT_EQ(ReadFilterStatus::StopIteration, filter_->onNewSession());
+  EXPECT_EQ(ReadFilterStatus::StopIteration, filter_->onData(recv_data_stub1_));
+
+  EXPECT_CALL(callbacks_, continueFilterChain());
+  filter_->onLoadDnsCacheComplete(
+      std::make_shared<Extensions::Common::DynamicForwardProxy::MockDnsHostInfo>());
+  EXPECT_CALL(*handle, onDestroy());
+}
+
+TEST_F(DynamicProxyFilterTest, LoadingCacheEntryWithDefaultBufferConfig) {
+  FilterConfig config;
+  config.mutable_buffer_options();
+  setup(config);
+
+  setFilterState("host", 50);
+  EXPECT_TRUE(filter_config_->bufferEnabled());
+  Upstream::ResourceAutoIncDec* circuit_breakers_{
+      new Upstream::ResourceAutoIncDec(pending_requests_)};
+  EXPECT_CALL(*dns_cache_manager_->dns_cache_, canCreateDnsRequest_())
+      .WillOnce(Return(circuit_breakers_));
+  Extensions::Common::DynamicForwardProxy::MockLoadDnsCacheEntryHandle* handle =
+      new Extensions::Common::DynamicForwardProxy::MockLoadDnsCacheEntryHandle();
+  EXPECT_CALL(*dns_cache_manager_->dns_cache_, loadDnsCacheEntry_(Eq("host"), 50, _, _))
+      .WillOnce(Return(
+          MockLoadDnsCacheEntryResult{LoadDnsCacheEntryStatus::Loading, handle, absl::nullopt}));
+  EXPECT_EQ(ReadFilterStatus::StopIteration, filter_->onNewSession());
+  // Two datagrams will be buffered.
+  EXPECT_EQ(ReadFilterStatus::StopIteration, filter_->onData(recv_data_stub1_));
+  EXPECT_EQ(ReadFilterStatus::StopIteration, filter_->onData(recv_data_stub2_));
+
+  EXPECT_CALL(callbacks_, continueFilterChain());
+  EXPECT_CALL(callbacks_, injectDatagramToFilterChain(_)).Times(2);
+  filter_->onLoadDnsCacheComplete(
+      std::make_shared<Extensions::Common::DynamicForwardProxy::MockDnsHostInfo>());
+  EXPECT_CALL(*handle, onDestroy());
+  EXPECT_FALSE(filter_config_->bufferEnabled());
+}
+
+TEST_F(DynamicProxyFilterTest, LoadingCacheEntryWithBufferSizeOverflow) {
```

Comment:
Might be worth validating the stat-write in this test, I don't think it's validated anywhere else.

Reply:
I think it may be hard testing it in a unit test, but it's validated in the integration test

Reply:
Ah, you're right, I missed that (I thought I'd seen it, but when I searched for 'stat' in the test files it wasn't there - guess I needed to look manually). Thanks!